runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -115,55 +111,60 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
-          @Override
-          public Optional<Boolean> call() throws Exception {
-            try {
-              for (;;) {
-                JobMetrics metrics = getJobMetrics(job);
-                Optional<Boolean> success = checkForPAssertSuccess(job, metrics);
-                if (messageHandler.hasSeenError()) {
-                  return Optional.of(false);
-                }
-
-                if (success.isPresent() && (!success.get() || atMaxWatermark(job, metrics))) {
-                  // It's possible that the streaming pipeline doesn't use PAssert.
-                  // So checkForSuccess() will return true before job is finished.
-                  // atMaxWatermark() will handle this case.
-                  return success;
-                }
-                Thread.sleep(10000L);
-              }
-            } finally {
-              if (!job.getState().isTerminal()) {
-                LOG.info("Cancelling Dataflow job {}", job.getJobId());
-                job.cancel();
-              }
+        // In streaming, there are infinite retries, so rather than timeout
+        // we try to terminate early by polling and canceling if we see
+        // an error message
+        while (true) {
+          State state = job.waitUntilFinish(Duration.standardSeconds(3), messageHandler);
+          if (state != null && state.isTerminal()) {
+            break;
+          }
+
+          if (messageHandler.hasSeenError()) {
+            if (!job.getState().isTerminal()) {
+              LOG.info("Cancelling Dataflow job {}", job.getJobId());
+              job.cancel();
             }
+            break;
           }
-        });
+        }
+
+        // Whether we canceled or not, this gets the final state of the job or times out
         State finalState =
             job.waitUntilFinish(
                 Duration.standardSeconds(options.getTestTimeoutSeconds()), messageHandler);
+
+        // Getting the final state timed out; it may not indicate a failure.
+        // This cancellation may be the second
         if (finalState == null || finalState == State.RUNNING) {
           LOG.info(
               "Dataflow job {} took longer than {} seconds to complete, cancelling.",
               job.getJobId(),
               options.getTestTimeoutSeconds());
           job.cancel();
         }
-        success = resultFuture.get();
+
+        if (messageHandler.hasSeenError()) {
+          success = Optional.of(false);
+        }
       } else {
         job.waitUntilFinish(Duration.standardSeconds(-1), messageHandler);
-        success = checkForPAssertSuccess(job, getJobMetrics(job));
+        success = checkForPAssertSuccess(job);
       }
+
       if (!success.isPresent()) {
-        throw new IllegalStateException(
-            "The dataflow did not output a success or failure metric.");
+        if (options.isStreaming()) {
+          LOG.warn(
+              "The dataflow did not output a success or failure metric."
```

Comment:
Standardized on "Dataflow job _id_" since these appear in interleaved multithreaded output sometimes.